carta/cpp/core/Data/Image/DataSource.cpp

```diff
@@ -382,28 +354,57 @@ std::vector<std::pair<int,double> > DataSource::_getIntensityCache( int frameLow
     //Not all percentiles were in the cache.  We are going to have to look some up.
     if ( foundCount < percentileCount ){
 
-        std::vector < int > allIndices;
-        std::vector < double > allValues;
+        std::vector<std::pair<int,double> > allValues;
         int spectralIndex = Util::getAxisIndex( m_image, AxisInfo::KnownType::SPECTRAL );
-        _copyData( frameLow, frameHigh, spectralIndex, allIndices, allValues );
+
+        Carta::Lib::NdArray::RawViewInterface* rawData = _getRawData( frameLow, frameHigh, spectralIndex );
+        if ( rawData == nullptr ){
+            qCritical() << "Error: could not retrieve image data to calculate missing intensities.";
+            return intensities;
+        }
+        
+        Carta::Lib::NdArray::TypedView<double> view( rawData, false );
+
+        // read in all values from the view into an array
+        // we need our own copy because we'll do quickselect on it...
+    
+        // Preallocate space to avoid 
+        // running out of memory unnecessarily through dynamic allocation
+        std::vector<int> dims = rawData->dims();
+        int total_size = std::accumulate(dims.begin(), dims.end(), 1, std::multiplies<int>());
+        allValues.reserve(total_size);
+        int index = 0;
+
+        view.forEach( [&allValues, &index] ( const double  val ) {
+            if ( std::isfinite( val ) ) {
+                allValues.push_back( std::make_pair(index, val) );
+            }
+            index++;
+        }
+        );
+
         if ( allValues.size() > 0 ){
+
+            int divisor = total_size;
+            if (spectralIndex != -1) {
```

Comment:
`total_size` and `dims` are derived from `rawData`, which is a slice which already takes the frames into account. `dims[spectralIndex]` is already the correct number of frames.

`frameHigh - frameLow` would in fact be off by one because the slice includes the lower bound.